controllers/statuscheck/controller.go

```diff
@@ -0,0 +1,140 @@
+// Copyright Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package statuscheck
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/controllers/utils/recorder"
+)
+
+type Reconciler struct {
+	logger        logr.Logger
+	kubeClient    client.Client
+	eventRecorder recorder.ChaosRecorder
+	manager       Manager
+}
+
+func NewReconciler(logger logr.Logger, kubeClient client.Client, eventRecorder recorder.ChaosRecorder, manager Manager) *Reconciler {
+	return &Reconciler{
+		kubeClient:    kubeClient,
+		eventRecorder: eventRecorder,
+		logger:        logger,
+		manager:       manager,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	startTime := time.Now()
+	obj := &v1alpha1.StatusCheck{}
+	if err := r.kubeClient.Get(ctx, req.NamespacedName, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.logger.Info("status check is deleted")
+			// the StatusCheck is deleted, remove it from manger
+			r.manager.Delete(req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// if status check was completed previously, we don't want to redo the termination
+	if obj.IsCompleted() {
+		r.logger.Info("status check is already completed")
+		return ctrl.Result{}, nil
+	}
+
+	result, ok := r.manager.Get(*obj)
+	if !ok {
+		// if nil, add status check to manager
+		err := r.manager.Add(*obj)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		result, _ = r.manager.Get(*obj)
+	}
+
+	updateError := retry.RetryOnConflict(retry.DefaultBackoff, r.updateStatus(ctx, req, result, startTime))
```

Comment:
And the logic about `FailureThreshold` and `SuccessThreshold` appear twice, in worker `sameResultCount` and in `condition` `isThresholdExceed`

Reply:
> And the logic about `FailureThreshold` and `SuccessThreshold` appear twice, in worker `sameResultCount` and in condition `isThresholdExceed`

Yes, it really confuses me. Because at the beginning, I hope the controller to manage all conditions, and the worker only do simple executions but not complex logics.

> We could update the status of StatusCheck as soon as `worker` at the outside of this reconciliation, avoid store/cache any state in different places.

This gives me some inspiration! Do you mean let the worker manage the `SuccessThreshold` and  `FailureThreshold` conditions, and let the controller manager the `DurationExceed` and `Completed` conditions? I think it might work better.

Reply:
> This gives me some inspiration! Do you mean let the worker manage the `SuccessThreshold`
 and `FailureThreshold` conditions, and let the controller manager the `DurationExceed`
 and `Completed` conditions? I think it might work better.

A little different, the statuscheck manager would only update/append the ``StatusCheckRecord``, and the `controller-runtime` reconciler would update the conditions and other fields.

And we do not need resultCache anymore.

Another thing  is that we should resolve the leadership of `controller-manager` on statuscheck manager. When controller-manager lost the leadership, the statuscheck manager should also be closed/shutdown. But it seems not easy to do that... How about create a known issue?

Reply:
> Another thing is that we should resolve the leadership of controller-manager on statuscheck manager. When controller-manager lost the leadership, the statuscheck manager should also be closed/shutdown. But it seems not easy to do that... How about create a known issue?

Already created issue #3114 for this problem!